Enable test_api IMethodTest in OSS

alanwaketan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #62521

Summary:
This diff did the following few things to enable the tests:
1. Exposed IMethod as TORCH_API.
2. Linked torch_deploy to test_api if USE_DEPLOY == 1.

Test Plan:
./build/bin/test_api --gtest_filter=IMethodTest.*

To be noted, one needs to run `python torch/csrc/deploy/example/generate_examples.py` before the above command.

Differential Revision: [D30055372](https://our.internmc.facebook.com/intern/diff/D30055372)

Commit:
Enable test_api IMethodTest in OSS

Summary:
This diff did the following few things to enable the tests:
1. Exposed IMethod as TORCH_API.
2. Linked torch_deploy to test_api if USE_DEPLOY == 1.
3. Generated the corresponding deploy examples while compiling imethod.cpp.

Test Plan:
./build/bin/test_api --gtest_filter=IMethodTest.*

[ghstack-poisoned]

Commit:
Update on "Enable test_api IMethodTest in OSS"


Summary:
This diff did the following few things to enable the tests:
1. Exposed IMethod as TORCH_API.
2. Linked torch_deploy to test_api if USE_DEPLOY == 1.
3. Generated the corresponding deploy examples while compiling imethod.cpp.

Test Plan:
./build/bin/test_api --gtest_filter=IMethodTest.*

[ghstack-poisoned]

wconstab commented on test/cpp/api/CMakeLists.txt:
```diff
@@ -94,3 +102,12 @@ endif()
 target_include_directories(parallel_benchmark PRIVATE ${ATen_CPU_INCLUDE})
 target_link_libraries(parallel_benchmark PRIVATE torch)
+
+if(USE_DEPLOY)
```
so, you shouldn't have to do this since it is already being called by our CI in this script:

https://github.com/pytorch/pytorch/blob/master/.jenkins/pytorch/test.sh#L456

note, as I see this I am wondering why it was called in the CI script instead of cmake in the first place.  it might be better to move it here.  But since there might have been a reason for doing it that way and since it should already be available for you, i'd suggest doing that change in a separate PR if you want to do it.

wconstab replied:
oh, and if we do move the generate step into cmake, it should be done just once, in the torchdeploy CMakeLists rather than here.  We should still be able to get access to it somehow from here.

alanwaketan replied:
I noticed the CI script as well. Maybe that step is done in the CI script instead of the CMake rules is due to the CI failures that I'm having here.

The reason why I did it in the CMake is that if I'm running the test like: `./build/bin/test_api --gtest_filter=IMethodTest.*`, it doesn't seem to be able to trigger the CI script. Maybe I ran the test in a wrong way.

wconstab replied:
well the way it is set up now, you would have to run the generate step manually, it doesn't make sense to run the whole CI script manually.  so i favor moving it to CMakeLists.

alanwaketan replied:
Then let me spilt the patch into two:

1. Just includes the test without invoking the generate_examples.py.
2. Tries to move the generate_examples.py to CMakeLists.

wconstab replied:
yea, when you move it please try adding in one place (near test_deploy)

Commit:
Update on "Enable test_api IMethodTest in OSS"


Summary:
This diff did the following few things to enable the tests:
1. Exposed IMethod as TORCH_API.
2. Linked torch_deploy to test_api if USE_DEPLOY == 1.
3. Generated the corresponding deploy examples while compiling imethod.cpp.

Test Plan:
./build/bin/test_api --gtest_filter=IMethodTest.*

[ghstack-poisoned]

alanwaketan commented:
@alanwaketan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30055372).

wconstab commented on test/cpp/api/imethod.cpp:
```diff
@@ -8,30 +8,40 @@
+  return e ? e : path;
+}
+
+// TODO(jwtan): Considering moving the following tests to somewhere else as linking deploy (sub-classes)
```
i think it's ok as is for now.

but if you wanted to change it, I think maybe what I'd do is,
- always add imethod.cpp into TORCH_API_TEST_SOURCES
- use a preprocessor directive like `USE_DEPLOY` (idk if we even have one, but we could add one) to conditionally add the parts to the unit tests that use deploy. 

That way the scriptmethod part would still be tested always, and, i think logically the tests _do_ belong here with the other API tests

alanwaketan replied:
For the script method, this is definitely the right place. For the deploy, I guess I'm confused by the current state where it is a standalone library other than torch. Would it in the future be merged into the torch library? Or is it logically a part of the torch library but it requires special treatment to do the custom loader magic? (I'm okay to remove the comment.)

Commit:
Update on "Enable test_api IMethodTest in OSS"


Summary:
This diff did the following few things to enable the tests:
1. Exposed IMethod as TORCH_API.
2. Linked torch_deploy to test_api if USE_DEPLOY == 1.

Test Plan:
./build/bin/test_api --gtest_filter=IMethodTest.*

Differential Revision: [D30055372](https://our.internmc.facebook.com/intern/diff/D30055372)

[ghstack-poisoned]

alanwaketan commented:
@alanwaketan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30055372).

alanwaketan commented:
Thanks @wconstab for approving the diff.

Commit:
Update on "Enable test_api IMethodTest in OSS"


Summary:
This diff did the following few things to enable the tests:
1. Exposed IMethod as TORCH_API.
2. Linked torch_deploy to test_api if USE_DEPLOY == 1.

Test Plan:
./build/bin/test_api --gtest_filter=IMethodTest.*

To be noted, one needs to run `python torch/csrc/deploy/example/generate_examples.py` before the above command.

Differential Revision: [D30055372](https://our.internmc.facebook.com/intern/diff/D30055372)

[ghstack-poisoned]

alanwaketan commented:
@alanwaketan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30055372).
